File: day23_part2.py
# for part two, I realized
# realized the general problem was actually better with linked lists and
# a fast hash table lookup and rewrote it entirely.
import time
import array
from llist import dllist, dllistnode

input_str = '942387615'
def circular_next(dllist, node):
    if not node.next:
        return dllist.first
    else:
        return node.next

def remove_node(lst, node):
    lst.remove(node)
    return node

cups = dllist([int(x) for x in input_str] + list(range(10, 1000000+1)))
moves = 10000000

lookup = {}
x = cups.first
while x:
    lookup[x.value] = x
    x = x.next

min_cup = min(lookup.keys())
max_cup = max(lookup.keys())

current_cup = cups.first
for m in range(moves):
    picked_up = [remove_node(cups, circular_next(cups, current_cup)) for _ in range(3)]

    destination_cup = current_cup.value - 1
    while destination_cup in (x.value for x in picked_up) or destination_cup < min_cup:
        destination_cup -= 1
        if destination_cup < min_cup:
            destination_cup = max_cup

    destination_cup_node = lookup[destination_cup]
    for x in reversed(picked_up):
        cups.insertnode(x, circular_next(cups, destination_cup_node))

    current_cup = circular_next(cups, current_cup)

print(circular_next(cups, lookup[1]).value * circular_next(cups, circular_next(cups, lookup[1])).value)

File: day23_part2.c
#include <stdio.h>
#define UPPER_BOUND 1000001

int remove_next(int *next_arr, int n) {
  int i = next_arr[n];
  next_arr[n] = next_arr[next_arr[n]];
  return i;
}

int insert_next(int *next_arr, int node, int n) {
  next_arr[n] = next_arr[node];
  next_arr[node] = n;
}

int main() {
  int next_arr[UPPER_BOUND];
  int min_cup = 1;
  int max_cup = 1000000;
  const int input_list[] = {9, 4, 2, 3, 8, 7, 6, 1, 5};

  for(int i = 0; i < sizeof(input_list) / sizeof(int) - 1; i++) {
    next_arr[input_list[i]] = input_list[i+1];
  }
  next_arr[input_list[(sizeof(input_list) / sizeof(int)) - 1]] = 10;
  for(int i = 10; i < 1000000; i++) {
    next_arr[i] = i+1;
  }
  next_arr[UPPER_BOUND-1] = 9;

  int current_cup = 9;
  const int moves = 10000000;

  int picked_up[3];
  for(int m = 0; m < moves; m++) {
    for(int i = 0; i < 3; i++) {
      picked_up[i] = remove_next(next_arr, current_cup);
    }

    int dest_cup = current_cup - 1;
    while(1) {
      char in_picked_up = 0;
      for(int i = 0; i < 3; i++) {
        if(dest_cup == picked_up[i]) {
          in_picked_up = 1;
          break;
        }
      }
      if(!in_picked_up && dest_cup >= min_cup)
        break;

      dest_cup--;

      if(dest_cup < min_cup) {
        dest_cup = max_cup;
      }
    }

    for(int i = 2; i >= 0; i--) {
      insert_next(next_arr, dest_cup, picked_up[i]);
    }

    current_cup = next_arr[current_cup];
  }

  printf("%lu\n", ((unsigned long) next_arr[1]) * ((unsigned long) next_arr[next_arr[1]]));
  return 0;
}
